content/2020-08-08-how-does-di-work.md

```diff
@@ -0,0 +1,557 @@
+---
+title: How Does Ember's Dependency Injection System Work?
+image: /images/sweetgun-UrY8jgHHapw-unsplash.jpg
+imageMeta:
+  attribution: "甜心之枪 Sweetgun"
+  attributionLink: https://unsplash.com/photos/UrY8jgHHapw
+featured: true
+authors:
+  - nullvoxpopuli
+date: Sat Aug 08 2020 13:04:01 GMT-0400 (Eastern Daylight Time)
+tags:
+  - ember
+  - javascript
+  - architecture
+  - frontend
+---
+
+## Why?
+
+One of the most common things I hear from people who are new to Ember,
+new to programming in general, or coming from another frontend ecosystem
+(especially React and Vue), is that they think Ember's dependency injection
+system is too complicated and magical --
+too hard to reason about or know where the injected services come from.
+I, too, was in that boat -- until I really dove into how it works -- it was
+then that I began to understand why dependency injection even exists, and
+how it's actually _simpler_ than _not_ having it at all.
+
+But first, I'm a proponent of ["show, don't tell"](https://en.wikipedia.org/wiki/Show,_don%27t_tell),
+so we'll start by implementing dependency injection from scratch in order to demystify
+and have something concrete that can be referenced.
+
+## What is Dependency Injection?
+
+According to [Wikipedia](https://en.wikipedia.org/wiki/Dependency_injection)
+
+> *dependency injection* is a technique in which an object receives other obects that it depends on.
+
+_That's it_.
+
+So... this is dependency injection?
+
+```js
+let foo = new Foo()
+
+let bar = new Bar(foo);
+```
+
+yes!.
+
+The big deal with dependency injection usually comes from _managing_ how an object
+receives those other objects.
+
+
+## Why use Dependency Injection?
+
+For me personally, there are two reasons:
+1. Application State (data and functions) can be easily shared between components
+2. Testing is much easier and can be done in isolation
+
+For #1, there are many ways to share state between components, but I like that
+dependency injection provides a centralized pattern and location for that state
+as well as an ergonomic and light way to interact with that state.
+
+For #2, this is a little harder to boil down to a sentence or two, and ultimately
+comes down overall architecture of your app, how big your app is, and what sorts of
+things provide value when tested. For example, let's say you have some behavior
+for interacting with an external API, maybe it's the [Star Wars JSON api](https://stackoverflow.com/questions/14301389/why-does-one-use-dependency-injection),
```

Comment:
I was think the link was to show my the start wars json API, but its taking me to stackoverflow

Reply:
thanks!